src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs

```diff
@@ -7728,44 +7728,30 @@ private static string GetNamespaceToRemove(CompletionContext context, TypeComple
 
         internal static List<CompletionResult> CompleteHelpTopics(CompletionContext context)
         {
-            var results = new List<CompletionResult>();
-            string userHelpDir = HelpUtils.GetUserHomeHelpSearchPath();
-            string appHelpDir = Utils.GetApplicationBase(Utils.DefaultPowerShellShellID);
-            string currentCulture = CultureInfo.CurrentCulture.Name;
-
-            //search for help files for the current culture + en-US as fallback
-            var searchPaths = new string[]
-            {
-                Path.Combine(userHelpDir, currentCulture),
-                Path.Combine(appHelpDir, currentCulture),
-                Path.Combine(userHelpDir, "en-US"),
-                Path.Combine(appHelpDir, "en-US")
-            }.Distinct();
+            List<CompletionResult> results = new();
 
-            string wordToComplete = context.WordToComplete + "*";
-            try
+            ArrayList helpProviders = context.ExecutionContext.HelpSystem.HelpProviders;
+            HelpFileHelpProvider helpFileProvider = null;
+            for (int i = helpProviders.Count - 1; i >= 0; i--)
             {
-                var wildcardPattern = WildcardPattern.Get(wordToComplete, WildcardOptions.IgnoreCase);
-
-                foreach (var dir in searchPaths)
+                if (helpProviders[i] is HelpFileHelpProvider provider)
                 {
-                    var currentDir = new DirectoryInfo(dir);
-                    if (currentDir.Exists)
-                    {
-                        foreach (var file in currentDir.EnumerateFiles("about_*.help.txt"))
-                        {
-                            if (wildcardPattern.IsMatch(file.Name))
-                            {
-                                string topicName = file.Name.Substring(0, file.Name.LastIndexOf(".help.txt"));
-                                results.Add(new CompletionResult(topicName));
-                            }
-                        }
-                    }
+                    helpFileProvider = provider;
+                    break;
                 }
             }
-            catch (Exception)
+
+            Collection<string> filesMatched = MUIFileSearcher.SearchFiles($"{context.WordToComplete}*.help.txt", helpFileProvider.GetExtendedSearchPaths());
```

Comment:
Formally helpFileProvider can still be null here.

Reply:
I don't mind adding a null check, but can it really be null in practice? In what scenario would the help providers not include a `HelpFileHelpProvider`?

Reply:
There is an idea to decouple components from monolithic PowerShell. HelpSystem is first candidate.

Reply:
In that case, how about we leave this as it is but we make another PR where we move this help file completer into a proper ArgumentCompleter attribute for `Get-Help`? I mean if we want to compartmentalize the help system, then we might as well do it right.
In general I think it was a mistake to add all the command specific completers in the core command completion code but I guess it's there due to legacy reasons as the ArgumentCompleter attribute wasn't there from the beginning.

Reply:
1. NRE is very bad. We must eliminate any possibility. This check should be added.
2. Yes, that's how it is historically. But before doing anything in this direction, it needs to be coordinated with MSFT team. I've had several attempts to decouple the Help System, but the team still hasn't found the resources to finish this work.